ur_controllers/src/gpio_controller.cpp

```diff
@@ -0,0 +1,313 @@
+  }
+
+  // analog io
+  for (size_t i = 0; i < 2; ++i)
+  {
+    config.names.emplace_back("gpio/standard_analog_output_" + std::to_string(i));
+  }
+
+  for (size_t i = 0; i < 2; ++i)
+  {
+    config.names.emplace_back("gpio/standard_analog_input_" + std::to_string(i));
+  }
+
+  for (size_t i = 0; i < 4; ++i)
+  {
+    config.names.emplace_back("gpio/analog_io_type_" + std::to_string(i));
+  }
+
+  // tool
+  config.names.emplace_back("gpio/tool_mode");
+  config.names.emplace_back("gpio/tool_output_voltage");
+  config.names.emplace_back("gpio/tool_output_current");
+  config.names.emplace_back("gpio/tool_temperature");
+
+  for (size_t i = 0; i < 2; ++i)
+  {
+    config.names.emplace_back("gpio/tool_analog_input_" + std::to_string(i));
+  }
+  for (size_t i = 0; i < 2; ++i)
+  {
+    config.names.emplace_back("gpio/tool_analog_input_type_" + std::to_string(i));
+  }
+
+  // robot
+  config.names.emplace_back("gpio/robot_mode");
+  for (size_t i = 0; i < 4; ++i)
+  {
+    config.names.emplace_back("gpio/robot_status_bit_" + std::to_string(i));
+  }
+
+  // safety
+  config.names.emplace_back("gpio/safety_mode");
+  for (size_t i = 0; i < 11; ++i)
+  {
+    config.names.emplace_back("gpio/safety_status_bit_" + std::to_string(i));
+  }
+
+  return config;
+}
+
+controller_interface::return_type ur_controllers::GPIOController::init(const std::string& controller_name)
+{
+  initMsgs();
+
+  return ControllerInterface::init(controller_name);
+}
+
+controller_interface::return_type ur_controllers::GPIOController::update()
+{
+  publishIO();
+  publishToolData();
+  publishRobotMode();
+  publishSafetyMode();
+  return controller_interface::return_type::SUCCESS;
+}
+
+rclcpp_lifecycle::node_interfaces::LifecycleNodeInterface::CallbackReturn
+ur_controllers::GPIOController::on_configure(const rclcpp_lifecycle::State& previous_state)
+{
+  try
+  {
+    // register publisher
+    io_pub_ = get_node()->create_publisher<ur_msgs::msg::IOStates>("io_states", rclcpp::SystemDefaultsQoS());
+
+    tool_data_pub_ = get_node()->create_publisher<ur_msgs::msg::ToolDataMsg>("tool_data", rclcpp::SystemDefaultsQoS());
+
+    robot_mode_pub_ =
+        get_node()->create_publisher<ur_dashboard_msgs::msg::RobotMode>("robot_mode", rclcpp::SystemDefaultsQoS());
+
+    safety_mode_pub_ =
+        get_node()->create_publisher<ur_dashboard_msgs::msg::SafetyMode>("safety_mode", rclcpp::SystemDefaultsQoS());
+
+    set_io_srv_ = get_node()->create_service<ur_msgs::srv::SetIO>(
+        "set_speed_slider", std::bind(&GPIOController::setIO, this, std::placeholders::_1, std::placeholders::_2));
+
+    set_speed_slider_srv_ = get_node()->create_service<ur_msgs::srv::SetSpeedSliderFraction>(
+        "set_io", std::bind(&GPIOController::setSpeedSlider, this, std::placeholders::_1, std::placeholders::_2));
+  }
+  catch (...)
+  {
+    return LifecycleNodeInterface::CallbackReturn::ERROR;
+  }
+
+  return LifecycleNodeInterface::on_configure(previous_state);
```

Comment:
I think you should simply return:

```suggestion
  return LifecycleNodeInterface::CallbackReturn::SUCCESS;
```